serving/serving.go

```diff
@@ -82,13 +81,170 @@ func (serv *FeatureServer) TrainingData(req *pb.TrainingDataRequest, stream pb.F
 	return nil
 }
 
+func (serv *FeatureServer) TrainingTestSplit(stream pb.Feature_TrainingTestSplitServer) error {
+
+	var (
+		trainIter, testIter provider.TrainingSetIterator
+		isTrainFinished     bool
+		isTestFinished      bool
+	)
+
+	for {
+		if isTestFinished && isTrainFinished {
+			// If both iterators are finished, we can close the stream
+			serv.Logger.Infow("Both iterators are finished, closing stream")
+			return nil
+		}
+
+		req, err := stream.Recv()
+		if err != nil {
+			return err
+		}
+
+		id := req.GetId()
+		name, variant := id.GetName(), id.GetVersion()
+		logger := serv.Logger.With("Name", name, "Variant", variant, "RequestType", req.GetRequestType().String())
+
+		logger.Infow("Getting training test split", "Request", req.String())
+
+		featureObserver := serv.Metrics.BeginObservingTrainingServe(name, variant)
+		defer featureObserver.Finish()
+
+		switch req.GetRequestType() {
+		case pb.RequestType_INITIALIZE:
+			if err := serv.handleSplitInitializeRequest(stream, req, &trainIter, &testIter, logger); err != nil {
+				featureObserver.SetError()
+				return err
+			}
+		default:
+			if err := serv.handleSplitDataRequest(stream, req, &trainIter, &testIter, &isTestFinished, &isTrainFinished, logger); err != nil {
```

Comment:
it is kind of weird that all these variables are getting updated within this method especially for the isTestFinished and isTrainFinished